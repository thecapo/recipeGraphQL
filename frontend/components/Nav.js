import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import Signout from "./Signout";

const Nav = () => (
  <User>
    {({ data }) => {
      const me = data ? data.me : null;
      return (
        <NavStyles data-test="nav">
          <Link href="/">
            <a>Recipes</a>
          </Link>
          {me && (
            <>
              <Link href="/createRecipe">
                <a>Add Recipe</a>
              </Link>
              <Signout />
              <span>{`hello, ${me.name}`}</span>
            </>
          )}
          {!me && (
            <Link href="/signup">
              <a>Sign in</a>
            </Link>
          )}
        </NavStyles>
      );
    }}
  </User>
);

export default Nav;
