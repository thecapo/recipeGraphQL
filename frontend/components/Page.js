import React, { Component } from "react";
import styled, { createGlobalStyle } from "styled-components";
import Header from "./Header";
import Meta from "./Meta";

// theme can be added here

const StyledPage = styled.div`
  background: white;
  color: black;
`;

const Inner = styled.div`
  max-width: 1000px;
  margin: 0 auto;
  padding: 2rem;
`;

const GlobalStyle = createGlobalStyle`
  @font-face {
    font-family: 'radnika_next';
    src: url('/static/radnikanext-medium-webfont.woff2') format('woff2');
    font-weight: normal;
    font-style: normal;
  }
  html {
    box-sizing: border-box;
    font-size: 10px;
  }
  *, *:before, *:after {
    box-sizing: inherit;
  }
  body {
    padding: 0;
    margin: 0;
    font-size: 1.5rem;
    line-height: 2;
    font-family: 'radnika_next';
  }
  a {
    text-decoration: none;
    color: black;
  }
  button { font-family: 'radnika_next'; }
`;

class Page extends Component {
  render() {
    return (
      <StyledPage>
        <Meta />
        <Header />
        <GlobalStyle />
        <Inner>{this.props.children}</Inner>
      </StyledPage>
    );
  }
}

export default Page;
