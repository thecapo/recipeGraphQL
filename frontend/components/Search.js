import React, { Component } from "react";
import Downshift, { resetIdCounter } from "downshift";
import Router from "next/router";
import { ApolloConsumer } from "react-apollo";
import gql from "graphql-tag";
import debounce from "lodash.debounce";
import { DropDown, DropDownRecipe, SearchStyles } from "./styles/DropDown";

const SEARCH_ITEMS_QUERY = gql`
  query SEARCH_ITEMS_QUERY($searchTerm: String!) {
    recipes(where: { OR: [{ title_contains: $searchTerm }] }) {
      id
      title
      ingredients
      image
      largeImage
    }
  }
`;

function routeToRecipe(recipe) {
  Router.push({
    pathname: "/recipe",
    query: {
      id: recipe.id
    }
  });
}

class AutoComplete extends Component {
  state = {
    recipes: [],
    loading: false
  };
  onChange = debounce(async (e, client) => {
    console.log('Searching...');
    this.setState({ loading: true });
    const res = await client.query({
      query: SEARCH_ITEMS_QUERY,
      variables: { searchTerm: e.target.value }
    });
    this.setState({
      recipes: res.data.recipes,
      loading: false
    })
  }, 350)
  render() {
    resetIdCounter();
    return (
      <SearchStyles>
        <Downshift
          onChange={routeToRecipe}
          itemToString={item => (item === null ? "" : item.title)}
        >
          {({
            getInputProps,
            getItemProps,
            isOpen,
            inputValue,
            highlightedIndex
          }) => (
            <div>
              <ApolloConsumer>
                {client => (
                  <input
                    {...getInputProps({
                      type: "search",
                      placeholder: "Search your favorite recipe...",
                      id: "search",
                      className: this.state.loading ? "loading" : "",
                      onChange: e => {
                        e.persist();
                        this.onChange(e, client);
                      }
                    })}
                  />
                )}
              </ApolloConsumer>
              {isOpen && (
                <DropDown>
                  {this.state.recipes.map((item, index) => (
                    <DropDownRecipe
                      {...getItemProps({ item })}
                      key={item.id}
                      highlighted={index === highlightedIndex}
                    >
                      <img width="50" src={item.image} alt={item.title} />
                      {item.title}
                    </DropDownRecipe>
                  ))}
                  {!this.state.recipes.length && !this.state.loading && (
                    <DropDownRecipe>
                      No Recipe Found {inputValue}
                    </DropDownRecipe>
                  )}
                </DropDown>
              )}
            </div>
          )}
        </Downshift>
      </SearchStyles>
    );
  }
}

export default AutoComplete;
