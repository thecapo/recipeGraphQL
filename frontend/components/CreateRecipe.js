import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Router from "next/router";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { ALL_RECIPES_QUERY } from "./Recipes";

const CREATE_RECIPE_MUTATION = gql`
  mutation CREATE_RECIPE_MUTATION(
    $title: String!
    $ingredients: String!
    $instructions: String!
    $image: String
    $largeImage: String
  ) {
    createRecipe(
      title: $title
      ingredients: $ingredients
      instructions: $instructions
      image: $image
      largeImage: $largeImage
    ) {
      id
      title
    }
  }
`;

class CreateRecipe extends Component {
  state = {
    title: "",
    instructions: "",
    ingredients: "",
    image: "",
    largeImage: ""
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'recipeapp');
    const res = await fetch("https://api.cloudinary.com/v1_1/cawo/image/upload", {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    console.log(file)
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    })
  }

  render() {
    return (
      <Mutation
        mutation={CREATE_RECIPE_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: ALL_RECIPES_QUERY }]}
      >
        {(createRecipe, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createRecipe();
              Router.push({
                pathname: "/recipe",
                query: { id: res.data.createRecipe.id }
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  // required
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                  <img src={this.state.image} alt="Upload Preview" />
                )}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  defaultValue={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="ingredients">
                Ingredients
                <input
                  type="text"
                  id="ingredients"
                  name="ingredients"
                  placeholder="Ingredients..."
                  required
                  defaultValue={this.state.ingredients}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="instructions">
                Instructions
                <textarea
                  id="instructions"
                  name="instructions"
                  placeholder="Instructions..."
                  required
                  defaultValue={this.state.instructions}
                  onChange={this.handleChange}
                />
              </label>
              <button type="sumbit">Create</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateRecipe;
export { CREATE_RECIPE_MUTATION };