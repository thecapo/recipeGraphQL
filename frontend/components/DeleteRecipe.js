import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_RECIPES_QUERY } from './Recipes';

const DELETE_RECIPE_MUTATION = gql`
  mutation DELETE_RECIPE_MUTATION($id: ID!) {
    deleteRecipe(id: $id) {
      id
    }
  }
`;

class DeleteRecipe extends Component {
  update = (cache, payload) => {
    const data = cache.readQuery({ query: ALL_RECIPES_QUERY });
    data.recipes = data.recipes.filter(recipe => recipe.id !== payload.data.deleteRecipe.id);
    cache.writeQuery({ query: ALL_RECIPES_QUERY, data });
  };
  render() {
    return (
      <Mutation
        mutation={DELETE_RECIPE_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteRecipe, { error }) => (
          <button
            onClick={() => {
              if (confirm("Are you sure you want to delete this item?")) {
                deleteRecipe().catch(err => {
                  alert(err.message);
                });
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteRecipe;