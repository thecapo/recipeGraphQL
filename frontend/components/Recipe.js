import React, { Component } from "react";
import PropTypes from "prop-types";
import Link from "next/link";
import DeleteRecipe from "./DeleteRecipe";
import RecipeStyles from "../components/styles/RecipeStyles";
import Title from "../components/styles/Title";

class Recipe extends Component {
  static propTypes = {
    recipe: PropTypes.object.isRequired
  };

  render() {
    const { recipe } = this.props;
    const recipeSplit = recipe.ingredients.split(", ");
    return (
      <RecipeStyles>
        {recipe.image ? (
          <img src={recipe.image} alt={recipe.title} />
        ) : (
          <img src="../static/no-image.png" alt="no image" />
        )}
        <Title>
          <Link
            href={{
              pathname: "/recipe",
              query: { id: recipe.id }
            }}
          >
            <a>{recipe.title}</a>
          </Link>
        </Title>
        <ul>
          {recipeSplit.map((ingredient, index) => (
            <li key={index}>{ingredient}</li>
          ))}
        </ul>
        <div className="buttonList">
          <Link
            href={{
              pathname: "/recipe",
              query: { id: recipe.id }
            }}
          >
            <a>Read Recipe</a>
          </Link> 
          <Link
            href={{
              pathname: "/update",
              query: { id: recipe.id }
            }}
          >
            <a>Edit</a>
          </Link>
          <DeleteRecipe id={recipe.id}>Delete Recipe</DeleteRecipe>
        </div>
      </RecipeStyles>
    );
  }
}

export default Recipe;
