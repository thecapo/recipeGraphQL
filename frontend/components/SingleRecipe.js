import React, { Component } from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import Head from "next/head";
import Link from "next/link";
import DeleteRecipe from "./DeleteRecipe";
import styled from "styled-components";
import Error from "./ErrorMessage";

const SingleRecipeDiv = styled.div`
  img {
    display: block;
    margin-left: auto;
    margin-right: auto;
    width: 50%;
    padding-bottom: 50px;
  }
  p {
    text-align: center;
    font-size: 20px;
  }
  .buttonList {
    display: grid;
    text-align: center;
    width: 100%;
    border-top: 1px solid #e1e1e1;
    grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
    grid-gap: 1px;
    background: #e1e1e1;
    & > * {
      background: black;
      border: 0;
      font-family: "radnika_next";
      font-size: 1rem;
      padding: 0.3rem;
      text-decoration: none;
      color: white;
    }
  }
  .titleSingle {
    display: block;
    text-align: center;
    list-style: none;
    font-size: 30px;
    text-decoration-line: underline;
  }
`;

const SINGLE_RECIPE_QUERY = gql`
  query SINGLE_RECIPE_QUERY($id: ID!) {
    recipe(where: { id: $id }) {
      id
      title
      ingredients
      instructions
      image
      largeImage
      user {
        id 
        name
      }
    }
  }
`;

class SingleRecipe extends Component {
  render() {
    return (
      <Query
        query={SINGLE_RECIPE_QUERY}
        variables={{
          id: this.props.id
        }}
      >
        {({ data, loading, error }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.recipe) return <p>No recipe found...</p>;
          const recipe = data.recipe;
          const ingredientSplit = data.recipe.ingredients.split(", ");
          const instructionsSplit = data.recipe.instructions.split(", ");
          return (
            <React.Fragment>
              <Head>
                <title>MyRecipe App | {recipe.title}</title>
              </Head>
              <SingleRecipeDiv>
                {recipe.image ? (
                  <img
                    src={recipe.image}
                    alt={recipe.title}
                    className="center"
                  />
                ) : (
                  <img src="../static/no-image.png" alt="no image" />
                )}
                <div className="titleSingle">{recipe.title.toUpperCase()}</div>
                <ul className="ingredientCenter">
                  {ingredientSplit.map((ingredient, index) => (
                    <li key={index}>{ingredient}</li>
                  ))}
                </ul>
                <hr />
                <ol>
                  {instructionsSplit.map((instruction, index) => (
                    <li key={index}>{instruction}</li>
                  ))}
                </ol>
                <div className="buttonList">
                  <Link href="/">
                    <a>Back</a>
                  </Link>
                  <Link
                    href={{
                      pathname: "/update",
                      query: { id: recipe.id }
                    }}
                  >
                    <a>Edit</a>
                  </Link>
                  <DeleteRecipe id={recipe.id}>
                    Delete Recipe
                  </DeleteRecipe>
                </div>
              </SingleRecipeDiv>
            </React.Fragment>
          );
        }}
      </Query>
    );
  }
}

export default SingleRecipe;
export { SINGLE_RECIPE_QUERY };
