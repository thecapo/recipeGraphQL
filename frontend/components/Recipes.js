import React, { Component } from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import Recipe from "./Recipe";

const ALL_RECIPES_QUERY = gql`
  query ALL_RECIPES_QUERY {
    recipes {
      id
      title
      ingredients
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const RecipesList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: 1000px;
  margin: 0 auto;
  li {
   float: left;
   padding-right: 30px;
  }
`;

class Recipes extends Component {
  render() {
    return (
      <Center>
        <Query query={ALL_RECIPES_QUERY}>
          {({ data, error, loading }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <p>Error: {error.message}</p>;
            return (
              <RecipesList>
                {data.recipes.map(recipe => (
                  <Recipe recipe={recipe} key={recipe.id} />
                ))}
              </RecipesList>
            );
          }}
        </Query>
      </Center>
    );
  }
}

export default Recipes;
export { ALL_RECIPES_QUERY };
