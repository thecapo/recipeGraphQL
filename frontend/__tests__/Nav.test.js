import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
// import { fakeUser } from "../lib/testUtils";
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User'; 

const fakeUser = () => ({
  __typename: 'User',
  id: "1234",
  name: 'John Wick',
  email: 'email@email.com',
  permissions: ["ADMIN"]
});

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } }
  }
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } }
  }
];

// added the data-test="nav" in the nav.js
describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav/>
      </MockedProvider>
    )
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul');
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it('renders full nav when signed in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    )
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul');
    const span = wrapper.find('span')
    expect(nav.children().length).toBe(4);
    expect(nav.text()).toContain("Sign Out");
    expect(span.text()).toContain(`hello, ` + fakeUser().name);
  });
})