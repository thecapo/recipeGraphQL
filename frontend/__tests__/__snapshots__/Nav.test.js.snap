// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<ul
  className="NavStyles-x3z0n-0 cvWGBA"
  data-test="nav"
>
  <Link
    href="/"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Recipes
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign in
    </a>
  </Link>
</ul>
`;
